docs/source/rfc-configuration-overhaul.md

```diff
@@ -0,0 +1,893 @@
+
+Running Iroha with a config file that only contains `PUBLIC_KEY` and `PRIVATE_KEY` results in the following error:
+
+```
+Error:
+   0: Please add `p2p_addr` to the configuration.
+```
+
+This error message is misleading and inconsistent:
+
+- If the `p2p_addr` field is added to the config file, the error stays the same.
+- The `p2p_addr` field is not a root-level field, it's actually a part of the `TORII` configuration and should be
+  `TORII.P2P_ADDR`.
+- Error messages asking to add public and private keys use uppercase for configuration fields, while in this error the
+  field name is written in lowercase.
+
+#### 5. Config file with only `PUBLIC_KEY`, `PRIVATE_KEY`, and `TORII.P2P_ADDR` specified
+
+Running Iroha with a config file that only contains `PUBLIC_KEY`, `PRIVATE_KEY`, and `TORII.P2P_ADDR` results in the
+following error:
+
+```
+Error:
+   0: Please add `api_url` to the configuration.
+```
+
+Comparing this scenario to the previous one, we can notice that the `TORII.API_URL` was missing before as well but the
+previous error message didn't mention it.
+
+There are also two inconsistencies:
+
+- Similar to `p2p_addr` in the previous scenario, this field name is lowercase, while for public and private keys it was
+  uppercase.
+- While `TORII.P2P_ADDR` and `TORII.API_URL` share the same root-level (`TORII`), their names are different: "addr"
+  derived from "address" and "URL". Logically these are both addresses or URLs, and it would make sense for the names to
+  be aligned.
+
+#### 6. Config file with invalid `PUBLIC_KEY`
+
+Providing an invalid `PUBLIC_KEY` in the `config.json` results in the same error as when there is no config file at all,
+or an empty config file:
+
+```
+Error:
+   0: Please add `PUBLIC_KEY and PRIVATE_KEY` to the configuration.
+```
+
+This is not a helpful error message as it does not mention what exactly is wrong, why and where the parsing failed.
+
+#### 7. Invalid `IROHA2_PUBLIC_KEY` environment variable
+
+Providing an invalid `IROHA2_PUBLIC_KEY` environment variable results in the following error:
+
+```
+Error:
+   0: Failed to build configuration from env
+   1: Failed to deserialize the field `IROHA_PUBLIC_KEY`: JSON5: Key could not be parsed. Key could not be parsed. Invalid character 's' at position 4
+   2: JSON5: Key could not be parsed. Key could not be parsed. Invalid character 's' at position 4
+```
+
+While this message is more helful than the ones we discussed above, there are still issues:
+
+- The message contains repetition.
+- Without an input snippet, the `Invalid character 's' at position 4` part of the message is not as helpful as it could
+  be.
+
+#### 8. Config file with extra fields
+
+Providing a configuration file with extra fields that are not supposed to be there does not produce any errors at all.
+
+This might lead to a bad user experience when user expects some options to apply, but doesn't have any idea that those
```

Comment:
I am confused, because the purpose of default values is exactly to be applied when nothing is specified at all.

As for accidentally having an outdated ENV var, this RFC includes defining a deprecation policy. According to it, when an ENV var renaming takes place, the user will have a window between versions when they will get a warning/error for the deprecated variable before it is deleted entirely.

Reply:
In that case we should at least specify what fields have default values, I guess such a set should be as little as possible (i.e. log level etc only)

Reply:
Hmm, from internal discussions I remember the opposite request: we should have as many defaults as possible.

Anyway, on the configuration reference writing stage we will decide in details which fields we should have, what default values should be etc. At that stage we will also collect feedback from external users.

Reply:
In my opinion, both of you have a point.
I have a question regarding the deployment workflow (which may be covered by a part of the RFC and I'm missing something).
Let's assume a person who never used Iroha 2 performs the deployment.
How do they see the configuration parameters that were applied as defaults?
Maybe a verbose-running mode would help debug the deployment configurations at a later point.

Reply:
That's why we should just provide an example ('default') exhaustive configuration explicitly, but don't allow parameters not being provided with values

Reply:
> That's why we should just provide an example ('default') exhaustive configuration explicitly, but don't allow parameters not being provided with values

I respectfully strongly disagree. Because this makes any change in the configuration a breaking change for every customer instead of only for those that actually use the parameter. We want to minimize the possibility of breaking changes downstream and we want to commit to as little of API as possible

Let me give an example of how a configuration file should look like (I take inspiration from `mysql.conf`):
```toml
[sumeragi]
private_key = "pq239874294u42139384u243"
# block_time_ms: 2000
```

1. All possible parameters should be present in the configuration file
2. Parameters that have a default value **should be commented out** (but still present in the file)
3. If a parameter has a default value the default value **must be visible** in the config file

This way a user is aware of all possible parameter and what their default values are, even if they will only use a small subset of parameters

>  How do they see the configuration parameters that were applied as defaults?

@6r1d  configuration should be printed in the log or we should at least provide an option to do so. We can print additional debug information in the log such as the information was the parameter given a default value or was set by the user, was it set through a file or an environment variable, etc.

Reply:
@mversic the suggestion is definitely valid, though we still have to determine what parameters should not be allowed to have the default value, e.g. things like metadata size limit could be risky if a user starts a network without proper attention to the consequences of not configuring that specifically to their use case or constraints.
On the other hand, if someone migrates to a newer version of Iroha, updating the corresponding configuration to match that is not a decision making factor overhead (esp. compared to a blockstore migration).

To conclude, I'd not insist on my view, either choice is fine to me. If we allow the defaults we just should be much more careful about this for every parameter now and in the future.

Reply:
@Mingela, do I get you correctly: your point is that there are more parameters which should be provided by the user than we have now?

In other words, configuration like metadata size limits should be defined by the user anyway.

Reply:
we can always decide post factum on the set of parameters that will be required vs the ones that will have a default

Reply:
> do I get you correctly: your point is that there are more parameters which should be provided by the user than we have now?
> 
> In other words, configuration like metadata size limits should be defined by the user anyway.

@0x009922, I'm afraid I'm not entirely aware of all the parameters which have defaults, but in my opinion things like metadata limits should be described in the docs properly and left up to a user. 

I've just come to another idea to consider which is to introduce a config mode, i.e. `strict` vs `relaxed` (or like advanced/basic) based on which a user may omit specifying many parameters (prohibiting or enforcing defaults) keeping that decided by them still. On the other hand, we probably should not complicate things that much..

Reply:
> things like metadata limits should be described in the docs

Everything should no doubt be described in the docs!

> `strict` vs `relaxed`

It might be convenient. Let's leave this discussion to the stage of the actual documentation reference creation. It will be the best time to make decisions based on what might be the best for users.

Also, I had an idea of having `extends` mechanism. With it, users might _extend_ existing presets. For example:

```toml
extends = "./base.toml"
# or even some built-in preset?
extends = "built-in:strict"
```

`extends` mechanism plays well with the "Trace Configuration Resolution" section.